### PR TITLE
Decals/texture atlas

### PR DESCRIPTION
**Pull request by Paulm-Unity**

### CODE REVIEW

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs [hunk 1]
```diff
@@ -104,20 +90,60 @@ public Camera CurrentCamera
...
+
+        public class TextureScaleBias : IComparable
+        {
+            public Texture m_Texture = null;
+            public Vector4 m_ScaleBias = Vector4.zero;
+            public int CompareTo(object obj)
+            {
+                TextureScaleBias other = obj as TextureScaleBias;
+                int size = m_Texture.width * m_Texture.height;
+                int otherSide = other.m_Texture.width * other.m_Texture.height;
```
Comment (JulienIgnace-Unity) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs]: Typo otherSize?

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs [hunk 2]
```diff
@@ -525,15 +537,60 @@ public void RenderIntoDBuffer(CommandBuffer cmd)
...
         {
-			cmd.SetGlobalTexture(HDShaderIDs._DecalAtlasID, TextureAtlas.GetTexCache());
+            cmd.SetGlobalTexture(HDShaderIDs._DecalAtlas2DID, Atlas.AtlasTexture);
+        }
+
+        public void AddTexture(CommandBuffer cmd, TextureScaleBias textureScaleBias) //ref Vector4 scaleBias, Texture texture)
+        {
+            if (textureScaleBias.m_Texture != null)
+            {
+                textureScaleBias.m_ScaleBias = Atlas.AddTexture(cmd, textureScaleBias.m_Texture);
```
Comment (JulienIgnace-Unity) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs]: Would it be possible to pass textureScaleBias.m_ScaleBias as a ref to Atlas.AddTexture and make this function and this.AddTexture return bool instead? Seems a bit awkward to me to test for Zero vector and rely on a mutable member for this.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl [hunk 3]
```diff
@@ -65,40 +74,70 @@ void AddDecalContribution(PositionInputs posInput, inout SurfaceData surfaceData
...
-
-			if(diffuseIndex != -1)
+			positionDS = positionDS * float3(1.0, -1.0, 1.0) + float3(0.5, 0.0f, 0.5);	// decal clip space
+			if ((all(positionDS.xyz > 0.0f) && all(1.0f - positionDS.xyz > 0.0f)))
 			{
-				ApplyBlendDiffuse(DBuffer0, mask, positionDS.xz, diffuseIndex, DBUFFERHTILEBIT_DIFFUSE, decalBlend, lod);
-				alpha = alpha < decalBlend ? decalBlend : alpha;	// use decal alpha if it higher than transparent alpha
-			}
+				// clamp by half a texel to avoid sampling neighboring textures in the atlas
+				float2 clampAmount = float2(0.5f / _DecalAtlasResolution.x, 0.5f / _DecalAtlasResolution.y);
```
Comment (JulienIgnace-Unity) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: We should probably pass the 1.0f/_DecalAtlasResolution.xy as uniforms directly

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs (hunk 1 above)
Comment (Paulm-Unity) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs]: Done

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs (hunk 2 above)
Comment (Paulm-Unity) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Decal/DecalSystem.cs]: Done

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl (hunk 3 above)
Comment (Paulm-Unity) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: We could, but then in other places it is used directly, so we will still have to use 1/

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl [hunk 4]
```diff
@@ -14,36 +14,45 @@ DecalData FetchDecal(uint start, uint i)
...
-	float4 src = SAMPLE_TEXTURE2D_ARRAY_LOD(_DecalAtlas, sampler_DecalAtlas, texCoords, sliceIndex, lod);
+	float4 src = SAMPLE_TEXTURE2D_LOD(_DecalAtlas2D, _trilinear_clamp_sampler_DecalAtlas2D, texCoords, lod);
 	src.z = src.w;
 	src.w = blend;
 	dst.xyz = src.xyz * src.w + dst.xyz * (1.0f - src.w);
 	dst.w = dst.w * (1.0f - src.w);
 	matMask |= mapMask;
 }
 
+float ComputeTextureLOD(float2 uvdx, float2 uvdy, float2 scale)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: Move it to commonh.hlsl (in next PR)

Comment (sebastienlagarde) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Material/Decal/DecalUtilities.hlsl]: Done

**Pull request merged**